### PR TITLE
Added `SetCurrentValue`

### PR DESCRIPTION
**Pull request by TomEdwardsEnscape**
## What does the pull request do?
This change is a prerequisite for #9944. It introduces a direct analogue to WPF's [`DependencyObject.SetCurrentValue`](https://learn.microsoft.com/dotnet/api/system.windows.dependencyobject.setcurrentvalue) method, which sets the effective value of the property without changing its value source or detaching any bindings or style setters applied to it.

The existing `SetValue` method is sufficient when a property's value has `BindingPriority.LocalValue`; see the [`Local_Binding_Overwrites_Local_Value`](https://github.com/AvaloniaUI/Avalonia/blob/55ccecfa5045130395a756c9aa3bf85012945de8/tests/Avalonia.Base.UnitTests/AvaloniaObjectTests_Binding.cs#L674) test.

If the priority is anything else, or unknown, `SetCurrentValue` should be used instead. In particular, it should be used when a control is setting its own public `StyledProperty` values.

This commit only calls `SetCurrentValue` in one place: `TemplateBinding`. More uses will be added in future commits.

## What is the current behavior?
It is currently not possible to set a value of a `StyledProperty` without the danger of changing its value source.

Controls that want to set their own properties must instead use `DirectProperty`, which can only ever be set with `LocalValue` priority. But this is not ideal for properties that can also be set from styles or templates, and following recent changes `DirectProperty` has been prevented from receiving values from styles with conditional selectors (see #9800).

## What is the updated/expected behavior with this PR?
`SetCurrentValue` can be used instead of `SetValue` to safely update the current value of a `StyledProperty`, no matter where it came from. This enables controls to use `StyledProperty` in many more situations.

If the property is unset or has an inherited value when `SetCurrentValue` is called, an `EffectiveValue` is added with the new `Internal` priority. This is the lowest priority at which a value can be written.

`SetCurrentValue` can also be used with `DirectProperty`, but this call simply redirects to `SetValue`.

## How was the solution implemented (if it's not obvious)?
This change thankfully proved mostly additive. Calls to `SetCurrentValue` alter current behaviour when they reach `EffectiveValue<T>`, where the new value object is stored and returned by `EffectiveValue.Value` until something happens to change the effective value again.

A value object from `SetCurrentValue` is stored alongside the last "prioritized value". This is to support the following sequence of events:

1. A property value is set by a template
2. The control calls `SetCurrentValue` on that property
3. A style trigger becomes active and sets the property to a new value
4. The style trigger deactivates and the template value becomes effective again

The `SetCurrentValue` value object is discarded at step 3. At step 4, the original template value set from step 1 returns.

Additionally, some validation code has been moved into shared methods to avoid further duplication.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes
None

## Obsoletions / Deprecations
None

## Fixed issues
Does not directly fix any known issues, but unblocks #9944.

### CODE REVIEW

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028654-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs [hunk 1]
```diff
@@ -127,7 +181,7 @@ public void DisposeAndRaiseUnset(ValueStore owner, StyledPropertyBase<T> propert
 
             if (property.Inherits && owner.TryGetInheritedValue(property, out var i))
             {
-                oldValue = ((EffectiveValue<T>)i).Value;
+                oldValue = ((EffectiveValue<T>)i).LastPrioritizedValue;
```
Comment (tomenscape) [inline src/Avalonia.Base/PropertyStore/EffectiveValue`1.cs:184]: Many references to `LastPrioritizedValue` should be `ActiveValue` instead.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028672-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by grokys** [comment]
Before I jump in and review this PR, I'd like to see an example of where this is necessary. In [this comment](https://github.com/AvaloniaUI/Avalonia/issues/9944#issuecomment-1383103246) you say:

> The only problematic case I've noticed on master [is where a two-way TemplateBinding sets a value on the source object](https://github.com/AvaloniaUI/Avalonia/blob/1e72fb920c748275da1f7a822da343ef7f99fa0a/src/Markup/Avalonia.Markup/Data/TemplateBinding.cs#L109). The source property could easily be a StyledProperty with a binding of its own.

Could you give me a concrete example of where this problem arises?

**Comment by tomenscape** [comment]
I'm not aware of any current issues, because things only breaks when the property being set is a `StyledProperty` with a non-local value applied. As the PR states, it is merely a pre-requisite change for future work.

However, you can produce an error that this PR fixes by making a small change to `ScrollBar`. Just add this new property (or change `RangeBase.ValueProperty`, but that is more work):

```C#
public static readonly StyledProperty<double> ValueProperty =
    AvaloniaProperty.Register<ScrollBar, double>(nameof(Value));

public double Value
{
    get => GetValue(ValueProperty);
    set => SetValue(ValueProperty, value);
}
````

Now use the mouse wheel to scroll any `ScrollViewer`.

On master, the `ScrollBar` moves once but is thereafter stops responding to changes to the scroll offset. This happens because when the `Track` inside the `ScrollBar` template moves, a two-way `TemplateBinding` transfers the value of `Track.Offset` to `ScrollBar.Offset`.

On master the new value is set with local priority, which overrides and disconnects the outer binding between `ScrollBar.Offset` and `ScrollViewer.Offset`.

On this branch the new offset is set with `SetCurrentValue`, the binding persists, and `ScrollBar` continues synchronise its values.

**Comment by grokys** [comment]
I'm looking at this now, and mostly happy with it but I can't shake the feeling that it's a bit more complex than it needs to be... In particular the addition of `_currentValue` and `_hasCurrentValue` to `UncommonFields` and the addition of the `LastPrioritizedValue` and `ActiveValue` to `EffectiveValue` feels like it should be unnecessary? Also the `Internal` priority feels like it shouldn't be needed.

I'm probably overlooking something but it feels like setting the current value should just write to `EffectiveValue.Value` and that's it (obv. with coercion/property changed events etc). The complicating factor here would be calling `SetCurrentValue` on a value that's not set on the object but in that case could we just temporarily create an `EffectiveValue` with `Unset` or `Inherited` priority and the `IsOverriddenCurrentValue` flag set?

I'll continue to look into this and play with it but wanted to quickly get your thoughts.

**Comment by tomenscape** [comment]
Thanks for taking a look!

The complicating factor is that `EffectiveValue` is mutated too often. `ValueStore.ReevaluateEffectiveValue` comes along at various moments [and updates the current `EffectiveValue` with data from one of its `ValueFrame` objects](https://github.com/AvaloniaUI/Avalonia/blob/bf85f4d3211ed25e7b480e2f9bcc23f2f9d56d63/src/Avalonia.Base/PropertyStore/ValueStore.cs#L814).

We don't want to lose the `SetCurrentValue` object during a mere re-evaluation, but we _do_ want to update the underlying `SetValue`/`LastPrioritizedValue` object, in order to support the style trigger scenario I mentioned in the PR description. This requires that the two be stored separately.

> could we just temporarily create an EffectiveValue with Unset or Inherited priority and the IsOverriddenCurrentValue flag set?

We could do this, but I thought it was confusing to have the value be unset and yet not the default value, or to be inherited and yet not the same as the parent's value. Since Avalonia's `BindingPriority` is much more visible than WPF's `ValueSource`, I opted to diverge from Microsoft's design and create a new priority that expressed the situation more clearly.

I have no problem with switching your suggestion, if you would prefer that.

**Comment by grokys** [comment]
> The complicating factor is that EffectiveValue is mutated too often. ValueStore.ReevaluateEffectiveValue comes along at various moments and updates the current EffectiveValue with data from one of its ValueFrame objects

Ah yes that's probably the part I was missing! However in `EffectiveValue` we do have a reference to the current effective value so maybe that could be used to detect when the value has actually changed? Just thinking aloud really, not tried it.

> We could do this, but I thought it was confusing to have the value be unset and yet not the default value, or to be inherited and yet not the same as the parent's value.

Yeah possibly, but is it much more confusing than having an e.g. `Style` value that's not the same as the value from the actual style?

> Since Avalonia's BindingPriority is much more visible than WPF's ValueSource

Feels to me like a separate flag saying "this value is overridden" would be clear enough; we could also display that in devtools (e.g. "Priority: Unset (Overridden)" or something.

> I have no problem with switching your suggestion, if you would prefer that.

Great, I'm definitely not decided, just trying to understand how you got to the solution you ended up with really. I'll need to look at it some more and think about it a while I think.

**Comment by tomenscape** [comment]
> maybe that could be used to detect when the value has actually changed

We will always need separate storage for the two values due to the style trigger scenario, so I don't forsee changing/avoiding the `ReevaluateEffectiveValue` code path producing any benefit (for this PR).

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0028968-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by robloo** [comment]
Would be great to get this one in soon so we can continue faster on the direct property conversions.

**Comment by grokys** [comment]
Opened #10324 as an alternative implementation.

**Comment by grokys** [comment]
Closing as #10324 added this functionality.

**Pull request closed without merging**